File: src/main/scala/com/kirkwoodwest/scalawoods/LaunchPadMiniMK3Colors.scala
package com.kirkwoodwest.scalawoods

import com.kirkwoodwest.scalawoods.utils.ColorUtil

object LaunchPadMiniMK3Colors {
  val hexColorMap: List[String] = List(
    "333333", // 2
    "888888", // 3
    "AAAAAA", // 4
    "df73c7", // 5
    "e8100b", // 6
    "e12118", // 7
    "d13d31", // 8
    "bbb9ee", // 9
    "da7c39", // 10
    "ba7038", // 11
    "ad542d", // 12
    "cad0b6", // 13
    "bbde70", // 14
    "a8c85f", // 15
    "9db85e", // 16
    "86e1b6", // 17
    "58de6a", // 18
    "4eb655", // 19
    "46a350", // 20
    "3edcc5", // 21
    "02db73", // 22
    "08c75f", // 23
    "1fb85c", // 24
    "57e7e4", // 25
    "01d3a9", // 26
    "04af7d", // 27
    "08a260", // 28
    "3fe0de", // 29
    "00dbe8", // 30
    "01c2c2", // 31
    "0db1b3", // 32
    "4cdcf4", // 33
    "00d8f1", // 34
    "00a3d2", // 35
    "028db6", // 36
    "39cef7", // 37
    "00a7f6", // 38
    "007eec", // 39
    "0d6ccc", // 40
    "4fa2f4", // 41
    "006ff5", // 42
    "004be8", // 43
    "0143dc", // 44
    "1732ed", // 45
    "0001f0", // 46
    "0103ea", // 47
    "070de4", // 48
    "6777f6", // 49
    "2301f3", // 50
    "1b01eb", // 51
    "1804dc", // 52
    "c369f8", // 53
    "b000f5", // 54
    "8300ef", // 55
    "7c0be5", // 56
    "d271f1", // 57
    "e801e2", // 58
    "c60ccd", // 59
    "b40fcc", // 60
    "e93517", // 61
    "bd934c", // 62
    "b9ad56", // 63
    "9d9f62", // 65
    "3ab061", // 66
    "25a9b9", // 67
    "0a41ea", // 68
    "0202f2", // 69
    "1979c5", // 70
    "2c04f1", // 71
    "8f62e7", // 72
    "8866b6", // 73
    "e6172f", // 74
    "9bd0b6", // 75
    "9cc972", // 76
    "6fd371", // 77
    "39b962", // 78
    "03ccd4", // 79
    "0188f3", // 80
    "0148f5", // 81
    "2303f4", // 82
    "5c07f2", // 83
    "a841ed", // 84
    "945f44", // 85
    "d26143", // 86
    "88c66a", // 87
    "80d2a5", // 88
    "08e27c", // 89
    "58d4bf", // 90
    "5fc1dc", // 91
    "4abdee", // 92
    "537ff3", // 93
    "335ff1", // 94
    "5c5eed", // 95
    "9330f4", // 96
    "e60bf0", // 97
    "da824f", // 98
    "b8c374", // 99
    "8ed078", // 100
    "ada363", // 101
    "8f7f50", // 102
    "229d8b", // 103
    "46a0b7", // 104
    "5e62d9", // 105
    "3d57ec", // 106
    "ab8c9a", // 107
    "c9272c", // 108
    "cf75ae", // 109
    "ce756c", // 110
    "c2b694", // 111
    "aedab6", // 112
    "8cd695", // 113
    "5866d5", // 114
    "a9bfe2", // 115
    "72beeb", // 116
    "758ef6", // 117
    "6d6cf6", // 118
    "7588d8", // 119
    "8396eb", // 120
    "8db6f8", // 121
    "e1212e", // 122
    "c13135", // 123
    "38d26f", // 124
    "31a45f", // 125
    "c1c56f", // 126
    "b2aa61", // 127
    "d69a58" // 128
  )

  private val colorMap: Array[Array[Int]] = makeRGBColorMap

  // Builds color map for class...
  private def makeRGBColorMap: Array[Array[Int]] = {
    hexColorMap.map(ColorUtil.hex2Rgb).toArray
  }

  def getIndexBlack: Int = 0

  def getIndexFromRGB(R: Int, G: Int, B: Int): Int = {
    if (R == 0 && G == 0 && B == 0) return 0 // return 0 if black

    // Get distance from RGB
    var distance: Float = 900
    var shortestDistanceIndex = 0

    for ((color, i) <- colorMap.zipWithIndex) {
      val Array(r2, g2, b2) = color.map(_ / 255.0f)
      val c1 = Array(r2, g2, b2)
      val c2 = Array(R, G, B).map(_ / 255.0f)

      val newDistance: Float = ColorUtil.getDistance(c1(0), c1(1), c1(2), c2(0), c2(1), c2(2))
      if (newDistance < distance) {
        shortestDistanceIndex = i
        distance = newDistance
      }
    }

    shortestDistanceIndex + 1 // First color is skipped so we add one to the index
  }
}

File: src/main/java/com/kirkwoodwest/launchpadminimk3/hardware/LaunchPadMiniMK3Colors.java
package com.kirkwoodwest.launchpadminimk3.hardware;

import com.kirkwoodwest.openwoods.utils.ColorUtil;
import java.util.List;
import java.util.Arrays;

public class LaunchPadMiniMK3Colors {
    private static final List<String> hexColorMap = Arrays.asList(
        "333333", "888888", "AAAAAA", "df73c7", "e8100b", "e12118", "d13d31", "bbb9ee",
        "da7c39", "ba7038", "ad542d", "cad0b6", "bbde70", "a8c85f", "9db85e", "86e1b6",
        "58de6a", "4eb655", "46a350", "3edcc5", "02db73", "08c75f", "1fb85c", "57e7e4",
        "01d3a9", "04af7d", "08a260", "3fe0de", "00dbe8", "01c2c2", "0db1b3", "4cdcf4",
        "00d8f1", "00a3d2", "028db6", "39cef7", "00a7f6", "007eec", "0d6ccc", "4fa2f4",
        "006ff5", "004be8", "0143dc", "1732ed", "0001f0", "0103ea", "070de4", "6777f6",
        "2301f3", "1b01eb", "1804dc", "c369f8", "b000f5", "8300ef", "7c0be5", "d271f1",
        "e801e2", "c60ccd", "b40fcc", "e93517", "bd934c", "b9ad56", "9d9f62", "3ab061",
        "25a9b9", "0a41ea", "0202f2", "1979c5", "2c04f1", "8f62e7", "8866b6", "e6172f",
        "9bd0b6", "9cc972", "6fd371", "39b962", "03ccd4", "0188f3", "0148f5", "2303f4",
        "5c07f2", "a841ed", "945f44", "d26143", "88c66a", "80d2a5", "08e27c", "58d4bf",
        "5fc1dc", "4abdee", "537ff3", "335ff1", "5c5eed", "9330f4", "e60bf0", "da824f",
        "b8c374", "8ed078", "ada363", "8f7f50", "229d8b", "46a0b7", "5e62d9", "3d57ec",
        "ab8c9a", "c9272c", "cf75ae", "ce756c", "c2b694", "aedab6", "8cd695", "5866d5",
        "a9bfe2", "72beeb", "758ef6", "6d6cf6", "7588d8", "8396eb", "8db6f8", "e1212e",
        "c13135", "38d26f", "31a45f", "c1c56f", "b2aa61", "d69a58"
    );

    private static final int[][] colorMap = makeRGBColorMap();

    private static int[][] makeRGBColorMap() {
        int[][] map = new int[hexColorMap.size()][];
        for (int i = 0; i < hexColorMap.size(); i++) {
            map[i] = ColorUtil.hex2Rgb(hexColorMap.get(i));
        }
        return map;
    }

    public static int getIndexBlack() {
        return 0;
    }

    public static int getIndexFromRGB(int R, int G, int B) {
        if (R == 0 && G == 0 && B == 0) {
            return 0; // return 0 if black
        }

        float distance = 900;
        int shortestDistanceIndex = 0;

        for (int i = 0; i < colorMap.length; i++) {
            int[] color = colorMap[i];
            float r2 = color[0] / 255.0f;
            float g2 = color[1] / 255.0f;
            float b2 = color[2] / 255.0f;
            float[] c1 = {r2, g2, b2};
            float[] c2 = {R / 255.0f, G / 255.0f, B / 255.0f};

            float newDistance = ColorUtil.getDistance(c1[0], c1[1], c1[2], c2[0], c2[1], c2[2]);
            if (newDistance < distance) {
                shortestDistanceIndex = i;
                distance = newDistance;
            }
        }

        return shortestDistanceIndex + 1; // First color is skipped so we add one to the index
    }
}
